components/tidb_query_datatype/src/codec/mysql/duration.rs

```diff
@@ -230,16 +242,27 @@ mod parser {
                 let (rest, frac) = fraction(rest, fsp).ok()?;
 
                 if !rest.is_empty() {
-                    return None;
+                    if chars_len >= 12 {
+                        return None;
+                    }
+                    truncated_parse = true;
                 }
 
                 Some(Duration::new_from_parts(
                     neg, hhmmss[0], hhmmss[1], hhmmss[2], frac, fsp as i8,
                 ))
             });
 
+        // In order to keep compatible with TiDB, when input string can only be partially parsed by hhmmss_compact
+        // and it can match the datetime format, we fallback to parse it using datetime format.
+        if truncated_parse && fallback_to_datetime {
+            return hhmmss_datetime(ctx, rest, fsp).map_or(None, |(_, duration)| Some(duration));
+        }
+
         match duration {
-            Some(Ok(duration)) => Some(duration),
+            Some(Ok(duration)) => ctx
+                .handle_truncate(truncated_parse)
+                .map_or(Some(duration), |_| Some(duration)),
```

Comment:
A little bit strange. How about
```suggestion
            Some(Ok(duration)) => {
                ctx.handle_truncate(truncated_parse);
                Some(duration)
            }
```

Reply:
This is what I wrote at first. Unfortunately, compiler doesn't like this and will complain by warning since the return value of handle_truncate is result and does not be handled.

```rust
warning: unused `std::result::Result` that must be used
   --> components/tidb_query_datatype/src/codec/mysql/duration.rs:264:17
    |
264 | /                 ctx
265 | |                 .handle_truncate(truncated_parse);
    | |__________________________________________________^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: this `Result` may be an `Err` variant, which should be handled
```